Alloc.GetRefs() return slice of length 0 when file doesn't exist

Roshan-Mehta opened the pull request:
### Changes
- Add `StatusCode` in `oTreeResponse`; if the major blobbers are returning not found or bad request, then we return a slice of length 0 instead of a consensus error.

### Fixes
- Fix https://github.com/0chain/gosdk/issues/862

### Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/gosdk/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

### Associated PRs (Link as appropriate):
- blobber:
- 0chain:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

Roshan-Mehta committed:
Alloc.GetRefs() return slice of length 0 when file doesn't exist

Roshan-Mehta committed:
Merge branch 'staging' into fix/alloc-get-ref

Roshan-Mehta committed:
Merge branch 'staging' into fix/alloc-get-ref

peterlimg commented on zboxcore/sdk/filerefsworker.go:
```diff
@@ -61,7 +62,11 @@ func (o *ObjectTreeRequest) GetRefs() (*ObjectTreeResult, error) {
 
+	found := false
 	for _, oTreeResponse := range oTreeResponses {
+		if oTreeResponse.statusCode != http.StatusBadRequest { //It should be http.StatusNotFound, but the blobber is returning 400, if the file not found
```
Collect errors and return the true major error. Check how it processed the errors in [ProcessCopy](https://github.com/0chain/gosdk/blob/a751ddada893441aa6b56a2b5baa060dc9a8409b/zboxcore/sdk/copyworker.go#L174)

peterlimg commented on zboxcore/sdk/filerefsworker.go:
```diff
@@ -130,6 +140,7 @@ func (o *ObjectTreeRequest) getFileRefs(oTR *oTreeResponse, bUrl string) {
 	})
 	if err != nil {
 		oTR.err = err
+		oTR.statusCode = respStatusCode
```
We should also set statusCode to `oTR` when request successfully.

peterlimg commented on zboxcore/sdk/filerefsworker.go:
```diff
@@ -106,13 +114,15 @@ func (o *ObjectTreeRequest) getFileRefs(oTR *oTreeResponse, bUrl string) {
 	}
 	oResult := ObjectTreeResult{}
 	ctx, cncl := context.WithTimeout(o.ctx, time.Second*30)
+	respStatusCode := 500
```
We should use 200 as default code and set it to `oTR` after calling `zboxutil.HttpDo()` successfully.

peterlimg commented on zboxcore/sdk/filerefsworker.go:
```diff
@@ -106,13 +114,15 @@ func (o *ObjectTreeRequest) getFileRefs(oTR *oTreeResponse, bUrl string) {
 		}
 		defer resp.Body.Close()
 		respBody, err := ioutil.ReadAll(resp.Body)
+		respStatusCode = resp.StatusCode
```
Move code one line above, we should process the error right after it is returned, please do not insert any code that is not necessary there.

Roshan-Mehta committed:
Address comment

1. Determine if the file not found by counting the major error status code
2. Small fixes

Roshan-Mehta committed:
Merge branch 'fix/alloc-get-ref' of https://github.com/0chain/gosdk into fix/alloc-get-ref

Roshan-Mehta committed:
Merge branch 'staging' into fix/alloc-get-ref

peterlimg commented on zboxcore/sdk/filerefsworker.go:
```diff
@@ -82,7 +85,10 @@ func (o *ObjectTreeRequest) GetRefs() (*ObjectTreeResult, error) {
 	}
-
+	
+	if zboxutil.MajorStatusCode(oTreeResponseStatusCodes) == http.StatusBadRequest { //It should be http.StatusNotFound, but the blobber is returning 400 if the file not found
```
Have you checked the responsed message, blobber should return 400 status code along with error messages. We need we need to what happened, 400 could be returned with different errors, and we should only return empty `ObjectTreeResult{}` when the error message indicates 'no record found' or similar error. Otherwise, we should return the major error.

Roshan-Mehta replied:
Yes, The response message from the blobber is `"400 Bad Request"` only, with no indication of "no file found."

peterlimg replied:
I checked the blobber endpoint `/v1/file/refs/{allocation}`, it should give error message. Anyway, I had a test to call the `GetRefs()` with none exist remote path, it will does show the `invalid path` error message. The main issue that you don't see the error message is because the `getFileRefs` method ignored it. I'll some commits, and you can test with my commits to see the error message returned from blobbers.

Roshan-Mehta replied:
@peterlimg  Yes, I have corrected now using the error msg (from your commit).

Roshan-Mehta committed:
Add test for allocation GetRefs

GetRefs should return slice of length 0 when file is absent.

peterlimg committed:
Return and log error message from blobbers on getting alloc refs

Roshan-Mehta committed:
Check file absent using error msg

Roshan-Mehta committed:
Fix typo

Roshan-Mehta committed:
Merge branch 'staging' into fix/alloc-get-ref

peterlimg reviewed (approved):
LGTM

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | master  |
| gosdk  | fix/alloc-get-ref  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |
